lib/charms/acme_client_operator/v0/acme_client.py

```diff
@@ -0,0 +1,147 @@
+"""# acme_client Library.
+This library is designed to enable developers to easily create new charms for implementations of the Acme protocol.
+This library contains all the logic necessary to get certificates from a certificate provider using the acme protocol.
+The constructor takes the following:
+- Reference to the parent charm (CharmBase)
+- The implementation specific command (str)
+- Additional configuration that contains environment variables (dict)
+## Getting Started
+To get started using the library, you just need to fetch the library using `charmcraft`.
+```shell
+cd some-charm
+charmcraft fetch-lib charms.acme_client_operator.v0.acme_client
+```
+Then, to initialise the library:
+```python
+from charms.acme_client_operator.v0.acme_client import AcmeClient
+from charms.observability_libs.v0.kubernetes_service_patch import KubernetesServicePatch
+from ops.charm import CharmBase
+from ops.main import main
+from charms.tls_certificates_interface.v1.tls_certificates import (  # type: ignore[import]
+    CertificateCreationRequestEvent,
+    TLSCertificatesProvidesV1,
+)
+class ExampleAcmeCharm(CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.tls_certificates = TLSCertificatesProvidesV1(self, "certificates")
+        self.framework.observe(
+            self.tls_certificates.on.certificate_creation_request,
+            self._acme_client_operator.on_certificate_creation_request,
+        )
+        lego_cmd = f"lego --email example@email.com --dns namecheap --domains example.com"
+        self._acme_client_operator = AcmeClient(self, lego_cmd, self.additional_config)
+```
+Charms that leverage this library also need to specify a `provides` relation in their
+`metadata.yaml` file. For example:
+```yaml
+provides:
+  certificates:
+    interface: tls-certificates
+```
+"""
+
+# The unique Charmhub library identifier, never change it
+LIBID = "b3c9913b68dc42b89dfd0e77ac57236d"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+import logging
+from typing import Dict
+
+from charms.tls_certificates_interface.v1.tls_certificates import (  # type: ignore[import]
+    CertificateCreationRequestEvent,
+)
+from cryptography import x509
+from cryptography.x509.oid import NameOID
+from ops.charm import CharmBase
+from ops.framework import Object
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.pebble import ExecError
+
+logger = logging.getLogger(__name__)
+
+LEGO_CERTS_PATH = "/tmp/.lego/certificates/"
+class AcmeClient(Object):
```

Comment:
As discussed earlier today, I think this should be inheriting `CharmBase`, and be used as an alternative `CharmBase` by the final charm. This would prevent circular references and would also let us do things that are usually not done by charm libraries, like setting the status.

Reply:
Fixed
For now in a new PR so we can compare designs.
https://github.com/ghislainbourgeois/lego_operator/pull/4
Let me know what do you think